yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -1045,7 +1045,8 @@ object Client extends Logging {
           s"in favor of the $CONF_SPARK_JAR configuration variable.")
       System.getenv(ENV_SPARK_JAR)
     } else {
-      SparkContext.jarOfClass(this.getClass).head
+      SparkContext.jarOfClass(this.getClass).getOrElse(throw new SparkException("Could not " +
+        "find jar with spark-yarn related code, make sure SPARK_PREPEND_CLASSES is not set"))
```

Comment:
If you want to be more helpful, you can also mention that the user can set `spark.yarn.jar` to force Spark to use a specific Spark archive instead of trying to figure it out from the classpath.